core/chapters/c12_dictionaries.py

```diff
@@ -0,0 +1,510 @@
+
+
+Anyway, this is a normal shop where things have one price.
+
+This normal shop has normal customers with normal shopping lists like `['apple', 'box', 'cat']`.
+
+And even though your customers have calculators in their pockets, they still expect you to add up all the prices
+yourself and tell them how much this will all cost, because that's what normal shops do.
+
+So let's write a function that does that. Complete the function below, particularly the line `price = ...`
+
+
+    __copyable__
+    __program_indented__
+
+    def program(self):
+        def total_cost(cart, prices):
+            result = 0
+            for item in cart:
+                price = ... 
+                result += price
+            return result
+
+        assert_equal(
+            total_cost(
+                ['apple', 'box', 'cat'],
+                {'apple': 2, 'box': 5, 'cat': 100, 'dog': 100},
+                ),
+            107,
+        )
+        """ 
+
+        def solution(self):
+            def total_cost(cart: Dict[str, str], prices: Dict[str, str]):
+                result = 0
+                for item in cart:
+                    price = prices[item]
+                    result += price
+                return result
+
+            return total_cost
+
+        tests = [
+            ((['apple', 'box', 'cat'],  {'apple': 2, 'box': 5, 'cat': 100, 'dog': 100}), 107)
+        ]
+
+        hints = "TODO"
+
+    class shopping_cart4(VerbatimStep):
+        """
+Perfect! You publish your website and start dreaming about how rich you're going to be.
+
+But soon you get a complaint from a customer who wants to buy 5 million dogs...and 2 boxes to put them in.
+
+Your website allows buying the same items several times, e.g. `total_cost(['box', 'box'], {...})` works,
+but they have to add each item one at a time, and for some reason this customer doesn't want to click
+'Add to Cart' 5 million times. People are so lazy!
+
+Here's the new code for you to fix:
+
+    __copyable__
+    __program_indented__
+        """
+
+        def program(self):
+            def total_cost(cart, quantities, prices):
+                result = 0
+                for item in cart:
+                    price = ...
+                    quantity = ...
+                    result += price * quantity
+                return result
+
+            assert_equal(
+                total_cost(
+                    ['dog', 'box'],
+                    {'dog': 5000000, 'box': 2},
+                    {'apple': 2, 'box': 5, 'cat': 100, 'dog': 100},
+                ),
+                500000010,
+            )
+
+
+    class shopping_cart5(VerbatimStep):
+        """
+We've added another parameter called `quantities` to `total_cost`.
+Now `cart` is still a list of strings, but it doesn't have any duplicates.
+`quantities` is a dictionary where the keys are the items in `cart` and the corresponding values are the quantity
+of that item that the customer wants to buy.
+
+
```

Comment:
This line here is meant to be the boundary between steps, not the "We've added another..." line above. That text is part of the same step. In https://demo.hedgedoc.org/OH78A7uwTDOjeMM4v0QFLg?edit these boundaries are represented by `----`. Similarly the students are not required to run the original DNA program.